### PR TITLE
Frustum/BBox/Distorsion...

### PR DESCRIPTION
**Pull request by mbredif**
@elemoine, this is a WIP PR to give you quick and dirty preview of the PCFRUSTUM and PCBOX3 architecture and functionalities I have in mind.
Could you please review it and give me feedback about how you could integrate that with your own developments ?

### CODE REVIEW

**Comment by elemoine** [comment]
@mbredif I changed this PR's base branch to `master`. I am in the process of removing `dev` as discussed in https://github.com/LI3DS/pointcloud/issues/55.

File: pgsql/pc_editor.c [hunk 1]
```diff
@@ -13,6 +13,11 @@
 #include "pc_pgsql.h"	   /* Common PgSQL support for our type */
 
 Datum pcpatch_rotate_quaternion(PG_FUNCTION_ARGS);
+Datum pcpatch_translate(PG_FUNCTION_ARGS);
+Datum pcpatch_affine(PG_FUNCTION_ARGS);
+Datum pcpoint_rotate_quaternion(PG_FUNCTION_ARGS);
+Datum pcpoint_translate(PG_FUNCTION_ARGS);
+Datum pcpoint_affine(PG_FUNCTION_ARGS);
```
Comment (elemoine) [inline pgsql/pc_editor.c]: Oops, I guess I need to create a PR for those missing declarations. At the same time, I wonder why those declarations are needed.

**Comment by elemoine** [comment]
That's a ton of impressive work! And I must say that it's beyond me right now :) I'd really like to understand what you want to achieve and what problems you aim to solve with this code. Maybe we should have so proper discussion about that.

File: pgsql/pc_editor.c (hunk 1 above)
Comment (mbredif) [reply pgsql/pc_editor.c]: the error I got : 
```
LI3DS/pointcloud/pgsql/pc_editor.c:198:7: warning: no previous prototype for ‘pcpatch_projective’ [-Wmissing-prototypes]
 Datum pcpatch_projective(PG_FUNCTION_ARGS)
       ^
```

Comment (elemoine) [reply pgsql/pc_editor.c]: Could you please help me reproduce the problem? Do you use autotools or cmake? Do you use specific flags? Thanks.

Comment (elemoine) [reply pgsql/pc_editor.c]: Anyway, I created https://github.com/LI3DS/pointcloud/pull/61 for this.

**Comment by mbredif** [comment]
ok, that is now rebased and more stable, ready for review / merge

**Comment by elemoine** [comment]
Would be good to remove "WIP" from the first commit.

File: lib/cunit/cu_pc_frustum.c [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+		1,0,1,1,1,
+
+		1,0,1,1,1,
+		0,1,0,0,0,
+		1,0,1,1,1,
+		1,0,1,1,1,
+		1,0,1,1,1,
+	};
+	for(mode=1; mode<5; ++mode)
+		for(i=0; i<N; ++i)
```
Comment (elemoine) [inline lib/cunit/cu_pc_frustum.c]: I am picky, but could we try respecting the styling used upstream?

```c
for ( mode = 1; mode < 5; ++mode )
{
    for ( i = 0; i < N; ++i)
    {
```

Feel free to reject me if you think I am too pedantic :)

File: pgsql/expected/pointcloud.out [hunk 3]
```diff
@@ -362,8 +362,8 @@ DELETE FROM pa_test_dim;
 INSERT INTO pa_test_dim (pa)
 SELECT PC_Patch(PC_MakePoint(3, ARRAY[x,y,z,intensity]))
 FROM (
- SELECT  
- -127+a/100.0 AS x, 
+ SELECT
+ -127+a/100.0 AS x,
```
Comment (elemoine) [inline pgsql/expected/pointcloud.out]: Do you mind removing this change? It's unrelated, and it should be fixed upstream.

File: pgsql/pc_editor.c [hunk 4]
```diff
@@ -13,6 +13,13 @@
 #include "pc_pgsql.h"	   /* Common PgSQL support for our type */
 
 Datum pcpatch_rotate_quaternion(PG_FUNCTION_ARGS);
+Datum pcpatch_translate(PG_FUNCTION_ARGS);
+Datum pcpatch_affine(PG_FUNCTION_ARGS);
+Datum pcpatch_projective(PG_FUNCTION_ARGS);
+Datum pcpoint_rotate_quaternion(PG_FUNCTION_ARGS);
+Datum pcpoint_translate(PG_FUNCTION_ARGS);
+Datum pcpoint_affine(PG_FUNCTION_ARGS);
+Datum pcpoint_projective(PG_FUNCTION_ARGS);
```
Comment (elemoine) [inline pgsql/pc_editor.c]: Please remove this. This is being fixed with a separate PR.

File: lib/pc_matrix.c [hunk 5]
```diff
@@ -27,20 +46,51 @@ pc_matrix_43_set(PCMAT43 mat,
...
+
+
+
+/**
+* Set a 4x4 matrix (4 columns, 4 lines)
+*/
+void
+pc_matrix_44_copy(PCMAT44 res, const PCMAT44 mat)
+{
+	memcpy(res,mat,16*sizeof(double));
```
Comment (elemoine) [inline lib/pc_matrix.c]: (styling) Spaces are missing.

File: lib/pc_matrix.c [hunk 6]
```diff
@@ -27,20 +46,51 @@ pc_matrix_43_set(PCMAT43 mat,
...
+}
+
+/**
+* Set a 3 vector
+*/
+void
+pc_vector_3_copy(PCVEC3 res, const PCVEC3 vec)
+{
+	memcpy(res,vec,3*sizeof(double));
```
Comment (elemoine) [inline lib/pc_matrix.c]: (styling) Spaces.

**Review by elemoine — CHANGES_REQUESTED** [review]
This looks great! I just made a few minor comments that you might want to consider.

File: pgsql/expected/pointcloud.out (hunk 3 above)
Comment (mbredif) [reply pgsql/expected/pointcloud.out]: ok

**Comment by mbredif** [comment]
Ok, I took your comments into account.
I propose to squash the commits into 3 more semantic ones so that you can merge them, ok ?
- matrix
- box
- frustum

**Comment by elemoine** [comment]
Yes, please squash commits as you see fit. Using three commits sounds very good to me. Please tell me if you need help with the squashing/rebasing operations.

**Comment by mbredif** [comment]
Commits are squashed and tests are passing !
ready to merge

**Review by elemoine — APPROVED** [review]

**Comment by elemoine** [comment]
Cool, feel free to merge if you think this is ready to go!

**Comment by elemoine** [comment]
Thank you for the huge work!

**Comment by mbredif** [comment]
Thanks, your turn now to do the reorganization :)

**Comment by elemoine** [comment]
I get a lot of warnings with this PR applied. I use the same compilation process and flags as Travis. Can you look into it?

I know I should have checked before we merge this.

**Comment by elemoine** [comment]
Tell me if you don't have time, and I'll do it. But I don't want to start the reorg with all these warnings.

**Comment by mbredif** [comment]
#62 should do it. Feel free to merge or further work on it if you still get warnings.

**Pull request merged**